File: easy/125_Valid_Palindrome.py
# A phrase is a palindrome if, after converting all uppercase letters into
# lowercase letters and removing all non-alphanumeric characters,
# it reads the same forward and backward. Alphanumeric characters include letters and numbers.
#
# Given a string s, return true if it is a palindrome, or false otherwise.

# class Solution:
#     def isPalindrome(self, s: str) -> bool:
#         new_str = ''
#         for el in s:
#             if el.isalpha() or el.isdigit():
#                 new_str += el.lower()
#         return True if new_str[::-1] == new_str else False


# решение через re
class Solution:
    def isPalindrome(self, s: str) -> bool:
        import re
        s = re.sub('[^0-9a-z]+', "", s.lower())
        return s == s[::-1]


a = Solution()
print(a.isPalindrome("A man, a plan, a canal: Panama"))

File: go/easy/125_Valid_Palindrome.go
package main

import (
	"fmt"
	"strings"
	"unicode"
)

// A phrase is a palindrome if, after converting all uppercase letters into lowercase
// letters and removing all non-alphanumeric characters, it reads the same forward and backward.
// Alphanumeric characters include letters and numbers.
// Given a string s, return true if it is a palindrome, or false otherwise.
//
// Example 1:
// Input: s = "A man, a plan, a canal: Panama"
// Output: true
// Explanation: "amanaplanacanalpanama" is a palindrome.
//
// Example 2:
// Input: s = "race a car"
// Output: false
// Explanation: "raceacar" is not a palindrome.
// Example 3:
//
// Input: s = " "
// Output: true
// Explanation: s is an empty string "" after removing non-alphanumeric characters.
// Since an empty string reads the same forward and backward, it is a palindrome.
func isPalindrome_(s string) bool {
	f := func(r rune) rune {
		if !unicode.IsLetter(r) && !unicode.IsNumber(r) {
			return -1
		}

		return unicode.ToLower(r)
	}
	s = strings.Map(f, s)

	for i := 0; i < len(s); i++ {
		if s[i] != s[len(s)-i-1] {
			return false
		}
	}

	return true
}

func main() {
	s1 := "A man, a plan, a canal: Panama"
	s2 := "race a car"
	fmt.Println(isPalindrome_(s1))
	fmt.Println(isPalindrome_(s2))
}
